### ISSUE TITLE
Implement Convolution CV Data Ingestion Pipeline

### ISSUE THREAD

**Issue by kjpou1** [kjpou1/convolution-patterns#1]
### 📝 **Task Description**

Build a modular data ingestion pipeline for the **Convolution CV** project to process and organize indicator chart pattern images for CNN-based classification.

---

### 📦 **Goals**

* Convert raw indicator chart images from a staging directory into a clean, structured dataset.
* Support TensorFlow/Keras-compatible loading.
* Enable reproducible, auditable, and stratified dataset creation for image classification tasks.

---

### 📁 **Output Directory Structure**

```
artifacts/
└── data/
    ├── raw/                          # Immutable copy of source images (optional)
    ├── processed/
    │   ├── train/pattern_type/*.png
    │   ├── val/pattern_type/*.png
    │   └── test/pattern_type/*.png
    └── metadata/
        └── pattern_metadata.csv
```

---

### ⚙️ **Pipeline Components**

* `RawStorageService`: Copies original images from staging to `raw/` (if enabled).
* `IngestionService`: Extracts image metadata (`instrument`, `pattern_type`, hash).
* `SplitterService`: Performs stratified train/val/test split by `pattern_type`.
* `StorageService`: Writes split images to `processed/` directory.
* `MetadataService`: Saves `pattern_metadata.csv` with fields:

  ```csv
  image,instrument,pattern_type,split,image_hash,class_label
  ```

---

### 🧩 **Key Features**

* ✅ `--label-mode` with `pattern_only` (default) or `instrument_specific`
* ✅ Stratified splitting by `pattern_type` (default 70/15/15, configurable)
* ✅ `--preserve-raw` flag to optionally snapshot source input
* ✅ Deterministic shuffling using `random_seed`
* ✅ Metadata tracking for each image (path, hash, split, class label)

---

### 🚀 **Example CLI**

```bash
python run.py ingest \
  --staging-dir ./lc_staging \
  --label-mode pattern_only \
  --split-ratios 70 15 15 \
  --preserve-raw true
```
